### ISSUE TITLE
Invalid URL for local file on 2.6.16

### ISSUE THREAD

**Issue by dev77-07** [peak3d/inputstream.adaptive#701]
Having an issue playing something that the m3u8 file is on a local directory, just started happening with 2.6.16, it gives an invalid url error in the logs.

```
2021-05-21 23:30:17.960 T:7996    DEBUG <general>: CPlayerCoreFactory::GetPlayers(C:\Users\media\KodiAutoHDR\portable_data\userdata\addon_data\plugin.video.discoveryplus\tmp\temp.m3u8)
2021-05-21 23:30:17.960 T:7996    DEBUG <general>: CPlayerSelectionRule::GetPlayers: considering rule: system rules
2021-05-21 23:30:17.960 T:7996    DEBUG <general>: CPlayerSelectionRule::GetPlayers: matches rule: system rules
2021-05-21 23:30:17.960 T:7996    DEBUG <general>: CPlayerSelectionRule::GetPlayers: considering rule: mms/udp
2021-05-21 23:30:17.960 T:7996    DEBUG <general>: CPlayerSelectionRule::GetPlayers: considering rule: lastfm/shout
2021-05-21 23:30:17.960 T:7996    DEBUG <general>: CPlayerSelectionRule::GetPlayers: considering rule: rtmp
2021-05-21 23:30:17.960 T:7996    DEBUG <general>: CPlayerSelectionRule::GetPlayers: considering rule: rtsp
2021-05-21 23:30:17.960 T:7996    DEBUG <general>: CPlayerSelectionRule::GetPlayers: considering rule: streams
2021-05-21 23:30:17.960 T:7996    DEBUG <general>: CPlayerSelectionRule::GetPlayers: considering rule: dvd
2021-05-21 23:30:17.960 T:7996    DEBUG <general>: CPlayerSelectionRule::GetPlayers: considering rule: discimage
2021-05-21 23:30:17.960 T:7996    DEBUG <general>: CPlayerSelectionRule::GetPlayers: considering rule: sdp/asf
2021-05-21 23:30:17.960 T:7996    DEBUG <general>: CPlayerSelectionRule::GetPlayers: considering rule: nsv
2021-05-21 23:30:17.960 T:7996    DEBUG <general>: CPlayerSelectionRule::GetPlayers: considering rule: radio
2021-05-21 23:30:17.960 T:7996    DEBUG <general>: CPlayerCoreFactory::GetPlayers: matched 0 rules with players
2021-05-21 23:30:17.960 T:7996    DEBUG <general>: CPlayerCoreFactory::GetPlayers: adding videodefaultplayer (VideoPlayer)
2021-05-21 23:30:17.960 T:7996    DEBUG <general>: CPlayerCoreFactory::GetPlayers: for video=1, audio=0
2021-05-21 23:30:17.960 T:7996    DEBUG <general>: CPlayerCoreFactory::GetPlayers: for video=1, audio=1
2021-05-21 23:30:17.960 T:7996    DEBUG <general>: CPlayerCoreFactory::GetPlayers: added 1 players
2021-05-21 23:30:17.961 T:7996    DEBUG <general>: Radio UECP (RDS) Processor - new CDVDRadioRDSData::CDVDRadioRDSData
2021-05-21 23:30:17.961 T:7996     INFO <general>: VideoPlayer::OpenFile: plugin://plugin.video.discoveryplus/?action=play&video_id=1159695&video_type=EPISODE
2021-05-21 23:30:17.961 T:13156   DEBUG <general>: Thread VideoPlayer start, auto delete: false
2021-05-21 23:30:17.961 T:7996    DEBUG <general>: CApplication::OnPlayBackStarted: CApplication::OnPlayBackStarted
2021-05-21 23:30:17.962 T:13156    INFO <general>: Creating InputStream
2021-05-21 23:30:17.962 T:17424   DEBUG <general>: dbiplus::MysqlDatabase::connect replacing configured host 10.77.0.252 with resolved host 10.77.0.252
2021-05-21 23:30:17.962 T:736     DEBUG <general>: Thread BackgroundLoader start, auto delete: false
2021-05-21 23:30:17.962 T:736     DEBUG <general>: dbiplus::MysqlDatabase::connect replacing configured host 10.77.0.252 with resolved host 10.77.0.252
2021-05-21 23:30:17.967 T:17424   DEBUG <general>: Skipped 1 duplicate messages..
2021-05-21 23:30:17.967 T:17424   DEBUG <general>: Loading settings for plugin://plugin.video.discoveryplus/?action=play&video_id=1159695&video_type=EPISODE
2021-05-21 23:30:17.970 T:7996    DEBUG <general>: CVideoGUIInfo::InitCurrentItem(plugin://plugin.video.discoveryplus/?action=play&video_id=1159695&video_type=EPISODE)
2021-05-21 23:30:17.970 T:7996    DEBUG <general>: CPlayerGUIInfo::InitCurrentItem(plugin://plugin.video.discoveryplus/?action=play&video_id=1159695&video_type=EPISODE)
2021-05-21 23:30:17.970 T:4132    DEBUG <general>: dbiplus::MysqlDatabase::connect replacing configured host 10.77.0.252 with resolved host 10.77.0.252
2021-05-21 23:30:17.972 T:13156   DEBUG <general>: ADDON: Dll Initializing - InputStream Adaptive
2021-05-21 23:30:17.972 T:13156   DEBUG <general>: SECTION:LoadDLL(C:\Users\media\KodiAutoHDR\portable_data\addons\inputstream.adaptive\inputstream.adaptive.dll)
2021-05-21 23:30:17.973 T:13156    INFO <general>: AddOnLog: inputstream.adaptive: SetVideoResolution (3840 x 2035)
2021-05-21 23:30:17.974 T:13156   DEBUG <general>: AddOnLog: inputstream.adaptive: Open()
2021-05-21 23:30:17.974 T:13156   DEBUG <general>: AddOnLog: inputstream.adaptive: found inputstream.adaptive.manifest_type: hls
2021-05-21 23:30:17.974 T:13156   DEBUG <general>: AddOnLog: inputstream.adaptive: Initial bandwidth: 376140049 
2021-05-21 23:30:17.975 T:13156   DEBUG <CSettingsManager>: requested setting (PRERELEASEFEATURES) was not found.
2021-05-21 23:30:17.975 T:13156   DEBUG <general>: AddOnLog: inputstream.adaptive: MAXRESOLUTION selected: 0 
2021-05-21 23:30:17.975 T:13156   DEBUG <general>: AddOnLog: inputstream.adaptive: MAXRESOLUTIONSECURE selected: 0 
2021-05-21 23:30:17.975 T:13156   DEBUG <general>: AddOnLog: inputstream.adaptive: STREAMSELECTION selected: 0 
2021-05-21 23:30:17.975 T:13156   DEBUG <general>: AddOnLog: inputstream.adaptive: FORCENONSECUREDECODER selected: 0 
2021-05-21 23:30:17.975 T:13156   FATAL <general>: AddOnLog: inputstream.adaptive: Invalid url: / expected ()
2021-05-21 23:30:17.975 T:13156   ERROR <general>: AddOnLog: inputstream.adaptive: Could not open / parse manifest (C:/Users/media/KodiAutoHDR/portable_data/userdata/addon_data/plugin.video.discoveryplus/tmp/temp.m3u8)
```

Rolled back to 2.6.15 and it works fine
```
2021-05-21 23:31:18.106 T:7996    DEBUG <general>: CPlayerCoreFactory::GetPlayers(C:\Users\media\KodiAutoHDR\portable_data\userdata\addon_data\plugin.video.discoveryplus\tmp\temp.m3u8)
2021-05-21 23:31:18.106 T:7996    DEBUG <general>: CPlayerSelectionRule::GetPlayers: considering rule: system rules
2021-05-21 23:31:18.106 T:7996    DEBUG <general>: CPlayerSelectionRule::GetPlayers: matches rule: system rules
2021-05-21 23:31:18.106 T:7996    DEBUG <general>: CPlayerSelectionRule::GetPlayers: considering rule: mms/udp
2021-05-21 23:31:18.106 T:7996    DEBUG <general>: CPlayerSelectionRule::GetPlayers: considering rule: lastfm/shout
2021-05-21 23:31:18.106 T:7996    DEBUG <general>: CPlayerSelectionRule::GetPlayers: considering rule: rtmp
2021-05-21 23:31:18.106 T:7996    DEBUG <general>: CPlayerSelectionRule::GetPlayers: considering rule: rtsp
2021-05-21 23:31:18.106 T:7996    DEBUG <general>: CPlayerSelectionRule::GetPlayers: considering rule: streams
2021-05-21 23:31:18.106 T:7996    DEBUG <general>: CPlayerSelectionRule::GetPlayers: considering rule: dvd
2021-05-21 23:31:18.106 T:7996    DEBUG <general>: CPlayerSelectionRule::GetPlayers: considering rule: discimage
2021-05-21 23:31:18.106 T:7996    DEBUG <general>: CPlayerSelectionRule::GetPlayers: considering rule: sdp/asf
2021-05-21 23:31:18.106 T:7996    DEBUG <general>: CPlayerSelectionRule::GetPlayers: considering rule: nsv
2021-05-21 23:31:18.106 T:7996    DEBUG <general>: CPlayerSelectionRule::GetPlayers: considering rule: radio
2021-05-21 23:31:18.106 T:7996    DEBUG <general>: CPlayerCoreFactory::GetPlayers: matched 0 rules with players
2021-05-21 23:31:18.106 T:7996    DEBUG <general>: CPlayerCoreFactory::GetPlayers: adding videodefaultplayer (VideoPlayer)
2021-05-21 23:31:18.106 T:7996    DEBUG <general>: CPlayerCoreFactory::GetPlayers: for video=1, audio=0
2021-05-21 23:31:18.106 T:7996    DEBUG <general>: CPlayerCoreFactory::GetPlayers: for video=1, audio=1
2021-05-21 23:31:18.106 T:7996    DEBUG <general>: CPlayerCoreFactory::GetPlayers: added 1 players
2021-05-21 23:31:18.107 T:7996    DEBUG <general>: Radio UECP (RDS) Processor - new CDVDRadioRDSData::CDVDRadioRDSData
2021-05-21 23:31:18.107 T:7996     INFO <general>: VideoPlayer::OpenFile: plugin://plugin.video.discoveryplus/?action=play&video_id=1159695&video_type=EPISODE
2021-05-21 23:31:18.107 T:11992   DEBUG <general>: Thread VideoPlayer start, auto delete: false
2021-05-21 23:31:18.107 T:7996    DEBUG <general>: CApplication::OnPlayBackStarted: CApplication::OnPlayBackStarted
2021-05-21 23:31:18.108 T:19424   DEBUG <general>: Thread BackgroundLoader start, auto delete: false
2021-05-21 23:31:18.108 T:3444    DEBUG <general>: dbiplus::MysqlDatabase::connect replacing configured host 10.77.0.252 with resolved host 10.77.0.252
2021-05-21 23:31:18.108 T:11992    INFO <general>: Creating InputStream
2021-05-21 23:31:18.108 T:19424   DEBUG <general>: dbiplus::MysqlDatabase::connect replacing configured host 10.77.0.252 with resolved host 10.77.0.252
2021-05-21 23:31:18.112 T:7996    DEBUG <general>: Skipped 1 duplicate messages..
2021-05-21 23:31:18.112 T:7996    DEBUG <general>: CVideoGUIInfo::InitCurrentItem(plugin://plugin.video.discoveryplus/?action=play&video_id=1159695&video_type=EPISODE)
2021-05-21 23:31:18.113 T:7996    DEBUG <general>: CPlayerGUIInfo::InitCurrentItem(plugin://plugin.video.discoveryplus/?action=play&video_id=1159695&video_type=EPISODE)
2021-05-21 23:31:18.113 T:4132    DEBUG <general>: dbiplus::MysqlDatabase::connect replacing configured host 10.77.0.252 with resolved host 10.77.0.252
2021-05-21 23:31:18.114 T:11992   DEBUG <general>: ADDON: Dll Initializing - InputStream Adaptive
2021-05-21 23:31:18.114 T:11992   DEBUG <general>: SECTION:LoadDLL(C:\Users\media\KodiAutoHDR\portable_data\addons\inputstream.adaptive\inputstream.adaptive.dll)
2021-05-21 23:31:18.115 T:11992    INFO <general>: AddOnLog: inputstream.adaptive: SetVideoResolution (3840 x 2035)
2021-05-21 23:31:18.115 T:11992   DEBUG <general>: AddOnLog: inputstream.adaptive: Open()
2021-05-21 23:31:18.115 T:11992   DEBUG <general>: AddOnLog: inputstream.adaptive: found inputstream.adaptive.manifest_type: hls
2021-05-21 23:31:18.115 T:11992   DEBUG <general>: AddOnLog: inputstream.adaptive: Initial bandwidth: 376140049 
2021-05-21 23:31:18.116 T:11992   DEBUG <general>: AddOnLog: inputstream.adaptive: MAXRESOLUTION selected: 0 
2021-05-21 23:31:18.116 T:11992   DEBUG <general>: AddOnLog: inputstream.adaptive: MAXRESOLUTIONSECURE selected: 0 
2021-05-21 23:31:18.116 T:11992   DEBUG <general>: AddOnLog: inputstream.adaptive: STREAMSELECTION selected: 0 
2021-05-21 23:31:18.116 T:11992   DEBUG <general>: AddOnLog: inputstream.adaptive: FORCENONSECUREDECODER selected: 0 
2021-05-21 23:31:18.116 T:11992   DEBUG <general>: AddOnLog: inputstream.adaptive: Effective URL 
2021-05-21 23:31:18.116 T:11992   DEBUG <general>: AddOnLog: inputstream.adaptive: Download C:/Users/media/KodiAutoHDR/portable_data/userdata/addon_data/plugin.video.discoveryplus/tmp/temp.m3u8 finished
2021-05-21 23:31:18.116 T:11992    INFO <general>: AddOnLog: inputstream.adaptive: Successfully parsed .mpd file. #Periods: 1, #Streams in first period: 2, Type: live, Download speed: 47017506.1546 Bytes/s
```

**Comment by gauntiii** [comment]
> Having an issue playing something that the m3u8 file is on a local directory, just started happening with 2.6.16, it gives an invalid url

I have the same issue also.

**Comment by matthuisman** [comment]
IA is not meant to work with local files.... so surprised it used too work.
eg. if the file uses a relative url - what would be the base url for it to calculate the final url?

A proxy is the correct approach.

Can you post the contents of the temp.m3u8?
maybe we can see if we can stop the hard-error and at least try to play it.

@glennguy 
What do you reckon?
This PR should fix it: https://github.com/xbmc/inputstream.adaptive/pull/702
I just don't want users relying on local files to actually work.
Any relative paths etc in the file will break it. 
I didn't even know curl fileopen worked with local paths. haha

**Comment by matthuisman** [comment]
@yeahme49 @gauntiii
Could you please try a build from here:
https://jenkins.kodi.tv/blue/organizations/jenkins/xbmc%2Finputstream.adaptive/detail/PR-702/7/artifacts

**Comment by gauntiii** [comment]
> @yeahme49 @gauntiii
> Could you please try a build from here:
> https://jenkins.kodi.tv/blue/organizations/jenkins/xbmc%2Finputstream.adaptive/detail/PR-702/2/artifacts

This fixed the issue for me, but to be clear I was getting the same error in the above post while trying to watch any content with the Discovery+ addon. 
"FATAL <general>: AddOnLog: inputstream.adaptive: Invalid url: / expected ()"

**Comment by matthuisman** [comment]
discovery add-on must be using local files instead of a proxy to serve it's customised files.
"/userdata/addon_data/plugin.video.discoveryplus/tmp/temp.m3u8"

**Comment by dev77-07** [comment]
discovery add-on uses local files for the US discovery+ because they put the wrong frame rate in the m3u8 file, so the add-on downloads the file, adjusts the frame-rate and uses that new file for playback.

surprised to hear IA isn't supposed to work for local files since I used this issue https://github.com/xbmc/inputstream.adaptive/issues/296 to know that the path in the add-on needed / instead of \ 

just tested the linked build and it works fine for me. if there is a chance of IA failing on local files again in the future I may need to look into how to implement a proxy that could adjust the frame-rate in the m3u8 instead of making a local file and submit a pr to the discovery+ add-on.

**Comment by matthuisman** [comment]
yeh... that was sort of a "hack" I guess

i just left a comment for the author:
https://forum.kodi.tv/showthread.php?tid=340055&pid=3038938#pid3038938

I think the PR to get back the same behaviour will get out pretty quick.
But yeh... might be worth using the proxy method to be safe

**Comment by matthuisman** [comment]
@yeahme49 
made a few more changes, could you please try:
https://jenkins.kodi.tv/blue/organizations/jenkins/xbmc%2Finputstream.adaptive/detail/PR-702/7/artifacts

**Comment by dev77-07** [comment]
Just tested latest version and it works as well.

Do you have any examples of addons that use a proxy? Hoping to have something to look at for an idea on how to write the proxy code.

**Comment by matthuisman** [comment]
@yeahme49 
looks like your PRS to move d+ to proxy were merged and released in version 1.5.15.
Are you able to close this issue?
(I decided against allowing local paths again just to remove again in 20)
